homeassistant/components/input_datetime/__init__.py

```diff
@@ -354,13 +362,35 @@ def unique_id(self) -> typing.Optional[str]:
         return self._config[CONF_ID]
 
     @callback
-    def async_set_datetime(self, date_val, time_val):
+    def async_set_datetime(self, date=None, time=None, datetime=None, timestamp=None):
         """Set a new date / time."""
-        if not date_val:
-            date_val = self._current_datetime.date()
-        if not time_val:
-            time_val = self._current_datetime.time()
-        self._current_datetime = datetime.datetime.combine(date_val, time_val)
+        if timestamp:
+            datetime = dt_util.as_local(dt_util.utc_from_timestamp(timestamp)).replace(
+                tzinfo=None
+            )
+
+        if datetime:
+            date = datetime.date()
+            time = datetime.time()
+
+        if not self.has_date:
+            date = None
+
+        if not self.has_time:
+            time = None
+
+        if not date and not time:
+            raise vol.Invalid("Nothing to set")
```

Comment:
Should we raise ValueError instead? That's our standard when raising errors in service handlers to show to users.

Reply:
Yeah I thought it was a bit weird too. I think this will get caught by the websocket API and tell user invalid format.

Reply:
I don't think that we should raise ValueError, as specific errors are better. `vol.Invalid` is what we tend to use for bad input.